lib/z3_mappings.ml

```diff
@@ -265,160 +279,146 @@ module Fresh = struct
         in
         op' e1 e2
-    end
-
-    module I32 :
-      Op_intf.S
-        with type v := int32
-         and type t := Z3.Expr.expr
-         and type unop := Types.I32.unop
-         and type binop := Types.I32.binop
-         and type relop := Types.I32.relop
-         and type cvtop := Types.I32.cvtop
-         and type triop := Types.I32.triop = struct
-      open Types.I32
-      open Z3
-      include Bv
-
-      let encode_val i = BitVector.mk_numeral ctx (Int32.to_string i) 32
-
-      let encode_cvtop op e =
-        let op' =
-          match op with
-          | WrapI64 -> BitVector.mk_extract ctx 31 0
-          | TruncSF32 -> fun f -> FloatingPoint.mk_to_sbv ctx rtz f 32
-          | TruncUF32 -> fun f -> FloatingPoint.mk_to_ubv ctx rtz f 32
-          | TruncSF64 -> fun f -> FloatingPoint.mk_to_sbv ctx rtz f 32
-          | TruncUF64 -> fun f -> FloatingPoint.mk_to_ubv ctx rtz f 32
-          | ReinterpretFloat -> FloatingPoint.mk_to_ieee_bv ctx
-          | ToBool -> encode_relop Ne (encode_val 0l)
-          | OfBool ->
-            fun e -> Boolean.mk_ite ctx e (encode_val 1l) (encode_val 0l)
-          | ExtendSI32 | ExtendUI32 -> assert false
-        in
-        op' e
-
-      let encode_triop _op _e1 _e2 _e3 = assert false
-    end
 
-    module I64 :
-      Op_intf.S
-        with type v := int64
-         and type t := Z3.Expr.expr
-         and type unop := Types.I64.unop
-         and type binop := Types.I64.binop
-         and type relop := Types.I64.relop
-         and type cvtop := Types.I64.cvtop
-         and type triop := Types.I64.triop = struct
-      open Types.I64
-      open Z3
-      include Bv
-
-      let encode_val i = BitVector.mk_numeral ctx (Int64.to_string i) 64
-
-      let encode_cvtop op e =
+      let encode_cvtop sz op e =
         let op' =
-          match op with
-          | ExtendSI32 -> BitVector.mk_sign_ext ctx 32
-          | ExtendUI32 -> BitVector.mk_zero_ext ctx 32
-          (* rounding towards zero (aka truncation) *)
-          | TruncSF32 -> fun f -> FloatingPoint.mk_to_sbv ctx rtz f 64
-          | TruncUF32 -> fun f -> FloatingPoint.mk_to_ubv ctx rtz f 64
-          | TruncSF64 -> fun f -> FloatingPoint.mk_to_sbv ctx rtz f 64
-          | TruncUF64 -> fun f -> FloatingPoint.mk_to_ubv ctx rtz f 64
-          | ReinterpretFloat -> FloatingPoint.mk_to_ieee_bv ctx
-          | ToBool -> encode_relop Ne (encode_val 0L)
-          | OfBool ->
-            fun e -> Boolean.mk_ite ctx e (encode_val 1L) (encode_val 0L)
-          | WrapI64 -> assert false
+          match sz with
+          | Ty.S32 -> (
+            match op with
+            | WrapI64 -> BitVector.mk_extract ctx 31 0
+            | TruncSF32 -> fun f -> FloatingPoint.mk_to_sbv ctx rtz f 32
+            | TruncUF32 -> fun f -> FloatingPoint.mk_to_ubv ctx rtz f 32
+            | TruncSF64 -> fun f -> FloatingPoint.mk_to_sbv ctx rtz f 32
+            | TruncUF64 -> fun f -> FloatingPoint.mk_to_ubv ctx rtz f 32
```

Comment:
These four cases could be merged in only two ?

Reply:
Thanks, yes we can merge all of these! :grinning: